src/Tablebot/Plugin/Netrunner/Custom.hs

```diff
@@ -0,0 +1,121 @@
+nrToBool (NrBool b) = Just b
+nrToBool _ = Nothing
+
+-- | @keys@ maps accepted card record keys to their parser functions.
+keys :: [(String, Maybe String -> Maybe NrData)]
+keys =
+  [ ("advancement", fmap $ NrInt . (read :: String -> Int)),
+    ("points", fmap $ NrInt . (read :: String -> Int)),
+    ("link", fmap $ NrInt . (read :: String -> Int)),
+    ("cost", fmap $ NrInt . (read :: String -> Int)),
+    ("faction", fmap $ NrText . pack),
+    ("influence", fmap $ NrInt . (read :: String -> Int)),
+    ("flavour", fmap $ NrText . pack),
+    ("maxInf", fmap $ NrInt . (read :: String -> Int)),
+    ("keywords", fmap $ NrText . pack),
+    ("minSize", fmap $ NrInt . (read :: String -> Int)),
+    ("strength", fmap $ NrInt . (read :: String -> Int)),
+    ("text", fmap $ NrText . pack),
+    ("title", fmap $ NrText . pack),
+    ("trash", fmap $ NrInt . (read :: String -> Int)),
+    ("type", fmap $ NrText . pack),
+    ("unique", fmap $ NrBool . fuzzyBool)
+  ]
+
+-- | @fuzzyBool@ maps a string to true or false (whichever is closest).
+fuzzyBool :: String -> Bool
+fuzzyBool = closestValue [("true", True), ("false", False)]
+
+-- | @fuzzyType@ maps a string to the closest card type.
+fuzzyType :: String -> String
+fuzzyType = closestMatch ["agenda", "asset", "event", "hardware", "ice", "identity", "operation", "program", "resource", "upgrade"]
+
+-- | @fuzzyFaction@ maps a string to the closest card faction by name.
+fuzzyFaction :: NrApi -> String -> String
+fuzzyFaction api query =
+  let names = unpack . Faction.name <$> factions api
+      codes = unpack . Faction.code <$> factions api
+   in closestValue (zip names codes) query
+
+-- | @customCard@ takes a set of parameters and does its best to turn that data
+-- into a card.
+customCard :: NrApi -> [(String, String)] -> Card
+customCard api cardData =
+  let params = fromList $ catMaybes $ readParam <$> cardData
+   in Card
+        { advancement_cost = nrToInt =<< lookup "advancement" params,
+          agenda_points = nrToInt =<< lookup "points" params,
+          base_link = nrToInt =<< lookup "link" params,
+          Card.code = Nothing,
+          cost = nrToInt =<< lookup "cost" params,
+          deck_limit = Nothing,
+          faction_code = nrToText =<< lookup "faction" params,
+          faction_cost = nrToInt =<< lookup "influence" params,
+          flavor = nrToText =<< lookup "flavour" params,
+          illustrator = Nothing,
+          influence_limit = nrToInt =<< lookup "maxInf" params,
+          keywords = nrToText =<< lookup "keywords" params,
+          minimum_deck_size = nrToInt =<< lookup "minSize" params,
+          pack_code = Nothing,
+          position = Nothing,
+          quantity = Nothing,
+          Card.side_code = Nothing,
+          strength = nrToInt =<< lookup "strength" params,
+          stripped_text = nrToText =<< lookup "text" params, -- TODO - strip this
+          stripped_title = nrToText =<< lookup "title" params, -- TODO - ditto
```

Comment:
How urgent are these TODOs?

Reply:
Not at all; the stripped field is required to interface with the nrdb api (unless I want to write my own JSON parser for it). All they do is take the text and title fields and remove newlines and such.

Reply:
Practically it doesn't even matter. If you find this comment chain while trying to work out what those lines are for, forget about them.